platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/RelevantMethods.java

```diff
@@ -31,91 +34,80 @@ public class RelevantMethods {
     final List<ServiceMethod> factories;
     final List<ServiceMethod> configurers;
 
-    RelevantMethods(List<Method> decorators, List<Method> factories, List<Method> configurers) {
-        this.decorators = toServiceMethodList(decorators);
-        this.factories = toServiceMethodList(factories);
-        this.configurers = toServiceMethodList(configurers);
-    }
-
-    private static List<ServiceMethod> toServiceMethodList(List<Method> methods) {
-        List<ServiceMethod> result = new ArrayList<ServiceMethod>(methods.size());
-        for (Method method : methods) {
-            result.add(SERVICE_METHOD_FACTORY.toServiceMethod(method));
-        }
-        return result;
+    private RelevantMethods(List<ServiceMethod> decorators, List<ServiceMethod> factories, List<ServiceMethod> configurers) {
+        this.decorators = decorators;
+        this.factories = factories;
+        this.configurers = configurers;
     }
 
-    public static RelevantMethods getMethods(Class<?> type) {
+    public static RelevantMethods getMethods(Class<? extends ServiceProvider> type) {
         RelevantMethods relevantMethods = METHODS_CACHE.get(type);
         if (relevantMethods == null) {
-            relevantMethods = buildRelevantMethods(type);
+            relevantMethods = new RelevantMethodsBuilder(type).build();
             METHODS_CACHE.putIfAbsent(type, relevantMethods);
         }
-
         return relevantMethods;
     }
 
-    private static RelevantMethods buildRelevantMethods(Class<?> type) {
-        RelevantMethodsBuilder builder = new RelevantMethodsBuilder(type);
-        RelevantMethods relevantMethods;
-        addDecoratorMethods(builder);
-        addFactoryMethods(builder);
-        addConfigureMethods(builder);
-        relevantMethods = builder.build();
-        return relevantMethods;
-    }
+    private static class RelevantMethodsBuilder {
```

Comment:
💭 I've inlined `RelevantMethodsBuilder` here, that's where most of these changes come from.